### PR TITLE
Move core to typescript

### PR DESCRIPTION
**Pull request by geovannimp**
Issue: https://github.com/beekeeper-studio/beekeeper-studio/issues/301

When I open the PR I had already started refactoring, as it was fun, here you have the complete refactor.

Based in @rathboma answer, I didn't touched any .vue file and I'm totaly open to modify the PR to remove any unwanted refactored file.

I placed some TODOs with "bugs" I found when doing the refactor, I think this already prove typescript benefits.

There is a lot more that could be refactored, but the PR would be much bigger and have a lot more conflicts.

Please, let me know what you guys think about and sorry for the PR size.

### CODE REVIEW

**Comment by rathboma** [comment]
Hey!

Wow, this must have been a lot of work.

Not 100% sure what I think about moving so much to typescript. I certainly see some benefits, especially for stuff in 'lib', but I do worry about the accessibility of the project to new contributors.

Let me noodle on it a few days.

**Comment by geovannimp** [comment]
No problem, take the time you want.

I did all refactor trying to keep closer to JS as possible, only adding types. So accessibility shouldn't be too hard.

It take some time, but was cool to understand the application architecture. This is something that I think typescript make more accessible because before the refactor I had no idea what type of "variable" was present at store state `connectionConfigs` for example.

**Comment by rathboma** [comment]
@geovannimp Ok. So I just had to solve a bug that would have been caught with typescript (broken auto-updates).

So... I was going to say 'not now', but dammit lets do it. I think it could add some stability and will let me make the db libraries a lot more class-based. 

Questions:
- Can you resolve the last couple of merge conflicts?
- Are there any other things need to be done before this can for real be merged?

**Comment by geovannimp** [comment]
- Can you resolve the last couple of merge conflicts?
	Of course, I will try to do it today or tomorrow.

- Are there any other things need to be done before this can for real be merged?
	- I tested some features and they are all working, but some test by the maintainer would make me more confident.
	- Could you take a look at the TODOs I placed at the code? Those are bugs I found that can be at master or something that I refactored wrong.

**Comment by geovannimp** [comment]
@rathboma Rebase is done

**Comment by rathboma** [comment]
cool, I'll look through the TODOS over the next few days

File: src/lib/menu/ClientMenuActionHandler.ts
Comment (rathboma) [inline src/lib/menu/ClientMenuActionHandler.ts:9]: Can you combine this with an interface shared by this and the NativeMenuActionHandler? The interface can go in common/

File: src/main.ts
Comment (rathboma) [inline src/main.ts:42]: this is fine

File: src/shims-tsx.d.ts
Comment (rathboma) [inline src/shims-tsx.d.ts:3]: What is this for?

File: src/store/index.ts
Comment (rathboma) [inline src/store/index.ts:247]: I've coded this to work with or without a table. If no table - it fetches all columns for all tables. So this TODO is fine

File: src/store/index.ts
Comment (rathboma) [inline src/store/index.ts:258]: All, good, thanks.

File: tsconfig.json
Comment (rathboma) [inline tsconfig.json:3]: I don't really know anything about this file, but some naive questions:

es2018 - this works in browsers?
Any thing I specifically need to know about here?

**Review by rathboma — CHANGES_REQUESTED** [review]
Hey, answered all your TODO questions. I think there are a few minor tweaks to make -- you ok making them?

I'm still a little nervous about all the extra stuff I'll need to learn, but excited to add some more stability to the library code.

File: src/common/appdb/Connection.ts
Comment (geovannimp) [reply src/common/appdb/Connection.ts:33]: I checked and passwords has been saved without encryption. I will change to subscribers and check if it resolves the issue, but maybe we will need some migration to encrypt already saved connections

Comment (geovannimp) [reply src/common/appdb/Connection.ts:33]: I had to do the migration to fix this, now passwords are being saved encrypted at the database. I will open another PR because its a "critical" bug and can be merged already without typescript.

Comment (geovannimp) [reply src/common/appdb/Connection.ts:33]: https://github.com/beekeeper-studio/beekeeper-studio/pull/322

File: src/common/appdb/models/saved_connection.ts
Comment (geovannimp) [reply src/common/appdb/models/saved_connection.ts:63]: `port?: number` allows only `number` and `undefined`, but in many places of the code is used `someVaraible = null` and I would need to change every place to `someVaraible = undefined` so I created this `Nullable<>` type to keep code standard.

In my option, refactor all the code to remove those `Nullable<>` and use `someVaraible?` should be the first thing to do after this being merged at master.

File: src/common/appdb/models/saved_connection.ts
Comment (geovannimp) [reply src/common/appdb/models/saved_connection.ts:97]: I moved, but still a TODO.

`if (!this._sshMode !== 'userpass') {`

`!this._sshMode` return a `boolean` and `'userpass'` is a `string`, so this code always return true.

File: src/shims-tsx.d.ts
Comment (geovannimp) [reply src/shims-tsx.d.ts:3]: This is create automatically when I convert the project to typescript. It can be removed I think (or moved to typings) but dont cause any harm.

File: src/store/index.ts
Comment (geovannimp) [reply src/store/index.ts:258]: But can I remove those lines?

Because `query` is never used.

File: tsconfig.json
Comment (geovannimp) [reply tsconfig.json:3]: I think we could use some lower target, but this is the default when converting the project to typescript and the code seems to be working. With a higher target we will have a transpiled JS code closest to typescript without crazy workarounds.

File: src/lib/menu/ClientMenuActionHandler.ts
Comment (geovannimp) [reply src/lib/menu/ClientMenuActionHandler.ts:9]: Do you mean something like this?

```
type IMenuActionHandler = IClientMenuActionHandler | INativeMenuActionHandler
```

But what is the diference between `IClientMenuActionHandler` and `INativeMenuActionHandler`?

File: src/common/appdb/Connection.ts
Comment (rathboma) [reply src/common/appdb/Connection.ts:33]: you rock!

**Comment by geovannimp** [comment]
Fixed almost all TODOs.

I commented every change request/doubt and I'm waiting answers or to be marked as resolved.

And I think you missed some TODOs, so I placed my nickname in all remaining TODOs to be easier to find.

File: src/common/appdb/models/saved_connection.ts
Comment (rathboma) [reply src/common/appdb/models/saved_connection.ts]: oh I see. Hmm I'll look into that.

File: src/store/index.ts
Comment (rathboma) [reply src/store/index.ts]: Yep!

File: src/common/appdb/models/saved_connection.ts
Comment (rathboma) [inline src/common/appdb/models/saved_connection.ts:208]: Yes, I think this should just be:

`if(this._sshMode !== 'userpass')`

IE - clear the SSH password if not authenticating with password

File: src/lib/menu/ClientMenuActionHandler.ts
Comment (rathboma) [reply src/lib/menu/ClientMenuActionHandler.ts:9]: Ah, yea I wasn't clear here. There's no difference! They should use the same interface

Both NativeMenuActionHandler and ClientMenuActionHandler have the same interface. So there should be 3 files:

```
common/.../IMenuActionHandler
app/.../ClientMenuActionHandler (implements IMenuActionHandler)
background/.../NativeMenuActionHandler (implements IMenuActionHandler)
```

File: src/store/index.ts
Comment (rathboma) [inline src/store/index.ts:188]: I think this is ok?

File: src/store/index.ts
Comment (rathboma) [inline src/store/index.ts:257]: Yes these lines can be removed.

**Comment by rathboma** [comment]
ok, added more comments to the TODOS

File: src/shims-tsx.d.ts
Comment (rathboma) [reply src/shims-tsx.d.ts:3]: kk, we can just leave it

File: src/common/appdb/models/saved_connection.ts
Comment (rathboma) [reply src/common/appdb/models/saved_connection.ts]: Yeah that shoudn't have the first `!`. Typo.

File: src/common/appdb/models/saved_connection.ts
Comment (rathboma) [reply src/common/appdb/models/saved_connection.ts:63]: I think typeorm is fine with it being either undefined or null and setting to null? Is the problem that we're using both undefined and null in different places?

File: tsconfig.json
Comment (MasterOdin) [reply tsconfig.json:3]: You could go as high as `es2019` given that beekeeper uses chromium 80 (through electron 8.2.5) and a minimum node version of 12. Unless you wanted to support a lower version of node or electron, there is no reason to not make this as high as possible, using a combination of https://kangax.github.io/compat-table/es2016plus/ and https://node.green/ to see what features the environment matches.

File: src/common/appdb/models/saved_connection.ts
Comment (geovannimp) [reply src/common/appdb/models/saved_connection.ts:208]: Ok, but with this the `else if` don't work.

If `_sshMode` is not different from `"userpass"`, that means that `_sshMode` is `"userpass"` and cant be `"keyfile"`.

Typescript give me the following erro: 

```
This condition will always return 'true' since the types '"userpass"' and '"keyfile"' have no overlap.ts(2367)
```

Btw, I changed the if/else to only else. Only confirm to me that I catch this right.

File: src/common/appdb/models/saved_connection.ts
Comment (geovannimp) [reply src/common/appdb/models/saved_connection.ts:63]: Is not a problem, but with undefined with we do things like someObject?.propertie and with null we can't do this.

And to make type correctly based in what typeorm accepts, the type should be `port?: Nullable<number>` and we would need to check every place by null and undefined. So is better to use only null or only undefined, and undefined is more JS friendly.

File: tsconfig.json
Comment (geovannimp) [reply tsconfig.json:3]: I would stay this value for now, based that this is the default. But after this being merged we can play with this value for tests and check if we have any gain with higher/lower values.

File: src/store/index.ts
Comment (geovannimp) [reply src/store/index.ts:188]: The problem is this:
```
if (server.db[dbName]) {
```

As dbName can be `undefined`, this can happen: `server.db[undefined]`.

It's actually working I think, but typescript complains about using `undefined` as key.

```
Type 'undefined' cannot be used as an index type.ts(2538)
```

We can keep this way for now, but would be good to refactor any time.

**Comment by geovannimp** [comment]
Cool, I fixed every change request I think.

Still 3 TODOs at `lib/db/client.ts`

File: src/lib/db/client.ts
Comment (rathboma) [inline src/lib/db/client.ts:322]: I don't use this function, it's a leftover, so this is fine

Comment (rathboma) [reply src/lib/db/client.ts:322]: IE - don't worry about this function at all. You could just remove it if you like.

File: src/lib/db/client.ts
Comment (rathboma) [inline src/lib/db/client.ts:187]: good catch. This is a leftover from a refactor. Totally fine to just omit this.

**Review by rathboma — COMMENTED** [review]
I think these are all fine to ignore honestly.

**Comment by rathboma** [comment]
ok, github hid that diff from me, so I didn't see them.

Added comments -- basically we can ignore those if we want.

**Comment by rathboma** [comment]
@geovannimp One merge conflick, then I think we're good to go \0/

**Comment by geovannimp** [comment]
All done! :D

**Pull request merged**
